docs/design/trimmed-assemblies.md

```diff
@@ -0,0 +1,212 @@
+</ItemGroup>
+```
+
+The above will opt `MyAssembly.dll` into trimming. Note that the ItemGroup should contain assembly names without an extension, similar to [`TrimmerRootAssembly`](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming-options#root-assemblies). Before .NET 6 this would have been done with a target:
+
+```xml
+<Target Name="ConfigureTrimming"
+        BeforeTargets="PrepareForILLink">
+  <ItemGroup>
+    <ManagedAssemblyToLink Condition="'%(Filename)' == 'MyAssembly'">
+      <IsTrimmable>true</IsTrimmable>
+    </ManagedAssemblyToLink>
+  </ItemGroup>
+</Target>
+```
+
+## Future evolution
+
+As the .NET ecosystem shifts to support trimming of more libraries, we will be able to rely more on the trim analysis warnings to provide correctness guarantees. Since these guarantees are the same regardless of the `TrimMode`, we expect SDKs to move to more aggressive trimming defaults.
+
+We expect that the .NET SDK will eventually set `TrimMode` to `link` instead of `copyused` as it does today. Long-term, we may even go as far as enabling trimming of all assemblies by default when using ILLink. Existing MSBuild `IsTrimmable` and `TrimMode` metadata will continue to make it possible for the developer to opt in or out of trimming. We could additionally provide further knobs to simplify controlling trimming behavior and opting out of such defaults.
+
+### `TrimAllAssemblies` global opt-in
+
+We could make it it easier to enabling trimming for all assemblies with a simple boolean. This would be equivalent to setting `IsTrimmable` to `true` on every assembly that is input to the linker. For example:
+
+```xml
+<PropertyGroup>
+  <TrimAllAssemblies>true</TrimAllAssemblies>
+</PropertyGroup>
+```
+
+could be used instead of
+
+```xml
+<Target Name="ConfigureTrimming"
+        BeforeTargets="PrepareForILLink">
+  <ItemGroup>
+    <ManagedAssemblyToLink>
+      <IsTrimmable>true</IsTrimmable>
+    </ManagedAssemblyToLink>
+  </ItemGroup>
+</Target>
+```
+
+This could be set by default in future SDKs, or it could be set by the developer in SDKs where it is not the default. We prefer not to introduce such a property at the moment, because it makes it too easy to enable this more "dangerous" behavior. We may consider adding this in the future when more of the .NET ecosystem has been made compatible with trimming.
+
+### `AssemblyMetadata("IsTrimmable", "False")`
+
+With more aggressive defaults, it could make sense to support an attribute opt-out via `[assembly: AssemblyMetadata("IsTrimmable", "False")]`. This would provide a way for developers to indicate that their assemblies should not be trimmed.
+
+We should also consider whether such an opt-out should prevent the linker from rewriting the attributed assembly. This could be useful as a way to preserve assemblies that have invariants which would be broken by rewriting, or which contain data that would be removed by the linker even with the `copy` action.
```

Comment:
> should prevent the linker from rewriting the attributed assembly

At a minimum, it needs to re-write the references, right? For example, removing `System.Runtime` and adding `System.Private.CoreLib`.

Reply:
That's true - my concern is that a developer might expect such an attribute to prevent modification. I rewrote this section to clarify what I meant.